### PR TITLE
Allow GOAWAY to be sent from handlers after the Http2MultiplexCodec

### PR DESCRIPTION
**Pull request by exell-christopher**
Motivation:
When a server is load-shedding or preparing to shut-down gracefully, it would be nice to be able to issue a GOAWAY so that existing requests can finish, but new requests are turned away. The Http2FrameCodec seems to have all the correct handling for this, but the Http2MultiplexCodec does not allow for anything other than an Http2StreamFrame to be written. As far as I can tell server developers have no way to send a GOAWAY without extending the Http2MultiplexCodec itself which doesn't seem ideal.

Modification:
Update the write() functionality of the the Http2MultiplexCodec to allow Http2GoAwayFrame and Http2StreamFrame to be written.

Added one test, but feel like there may need to be more, just not sure at which level they need to be added.

My read through of the spec indicates that this shouldn't violate any rules.

### CODE REVIEW

**Comment by normanmaurer** [comment]
LGTM... @Scottmitch @nmittler PTAL

**Comment by normanmaurer** [comment]
Also maybe @buchgr

File: codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java [hunk 1]
```diff
@@ -343,6 +343,31 @@ public void settingChannelOptsAndAttrsOnBootstrap() {
...
+        };
+
+        Http2StreamChannelBootstrap b = new Http2StreamChannelBootstrap();
+        b.parentChannel(parentChannel).handler(childChannelInitializer);
+        Channel childChannel = b.connect().channel();
+        assertTrue(childChannel.isActive());
+
+        Http2GoAwayFrame goAwayFrame = parentChannel.readOutbound();
+        assertNotNull(goAwayFrame);
+
```
Comment (Scottmitch) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java]: we need to call `goAwayFrame.release()` when we are done with `goAwayFrame`

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java [hunk 2]
```diff
@@ -369,27 +369,35 @@ protected void doClose() throws Exception {
...
+                Http2StreamFrame frame = (Http2StreamFrame) msg;
+                ChannelPromise promise = ctx.newPromise();
+                if (isStreamIdValid(frame.streamId())) {
+                    ReferenceCountUtil.release(frame);
+                    throw new IllegalArgumentException("Stream id must not be set on the frame. Was: "
+                        + frame.streamId());
+                }
+                if (!isStreamIdValid(streamId())) {
+                    if (!(frame instanceof Http2HeadersFrame)) {
+                        throw new IllegalArgumentException("The first frame must be a headers frame. Was: "
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java]: looks like we missed a call to `ReferenceCountUtil.release(frame);` previously ... can you add this?

File: codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java (hunk 1 above)
Comment (exell-christopher) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecTest.java]: Fixing.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java (hunk 2 above)
Comment (exell-christopher) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java]: Will do.

**Comment by exell-christopher** [comment]
Changes Committed.

**Comment by Scottmitch** [comment]
4.1 (52aecab)

Thanks @exell-christopher !

**Comment by ejona86** [comment]
FYI: I think this was previously possible by writing the GOAWAY to the Channel.parent().

**Pull request closed without merging**